### DISCUSSION TITLE
Feedback on new core APIs for Streams, Shell Commands, Spinners, Continuation Style

### DISCUSSION THREAD

**Discussion by dillonkearns** [dillonkearns/elm-pages#470] [General]
I would love any feedback before I wrap up the docs and release these new APIs!

Notable new features:

- You can capture `FileStats` for Glob matches
- `Glob.fromString`
- Spinners for shell scripts
- Run shell commands through `Script.command` and `Script.exec`
- New `BackendTask.Stream` API for creating and running/reading a pipeline of Stream's (goes through NodeJS's native Stream APIs)
- `BackendTask.Do` module for some helpers for using continuation-style in your scripts or `BackendTask` definitions

I haven't written up the docs for some of these yet, but I'm moving into that phase to do the final wrapup. Any and all feedback is welcome and appreciated!

## ADDED MODULES - MINOR ##

[`BackendTask.Do`](https://elm-doc-preview.netlify.app/BackendTask-Do?repo=dillonkearns%2Felm-pages&version=master)
[`BackendTask.Stream`](https://elm-doc-preview.netlify.app/BackendTask-Stream?repo=dillonkearns%2Felm-pages&version=master)
[`Pages.Script.Spinner`](https://elm-doc-preview.netlify.app/Pages-Script-Spinner?repo=dillonkearns%2Felm-pages&version=master)

### Added
[`do`](https://elm-doc-preview.netlify.app/BackendTask?repo=dillonkearns%2Felm-pages&version=master#do) :
    BackendTask.BackendTask error value
    -> BackendTask.BackendTask error ()
[`doEach`](https://elm-doc-preview.netlify.app/BackendTask?repo=dillonkearns%2Felm-pages&version=master#doEach) :
    List.List (BackendTask.BackendTask error ())
    -> BackendTask.BackendTask error ()
[`failIf`](https://elm-doc-preview.netlify.app/BackendTask?repo=dillonkearns%2Felm-pages&version=master#failIf) :
    Basics.Bool
    -> FatalError.FatalError
    -> BackendTask.BackendTask FatalError.FatalError ()
[`inDir`](https://elm-doc-preview.netlify.app/BackendTask?repo=dillonkearns%2Felm-pages&version=master#inDir) :
    String.String
    -> BackendTask.BackendTask error value
    -> BackendTask.BackendTask error value
[`quiet`](https://elm-doc-preview.netlify.app/BackendTask?repo=dillonkearns%2Felm-pages&version=master#quiet) :
    BackendTask.BackendTask error value
    -> BackendTask.BackendTask error value
[`sequence`](https://elm-doc-preview.netlify.app/BackendTask?repo=dillonkearns%2Felm-pages&version=master#sequence) :
    List.List (BackendTask.BackendTask error value)
    -> BackendTask.BackendTask error (List.List value)
[`withEnv`](https://elm-doc-preview.netlify.app/BackendTask?repo=dillonkearns%2Felm-pages&version=master#withEnv) :
    String.String
    -> String.String
    -> BackendTask.BackendTask error value
    -> BackendTask.BackendTask error value

### Added

[`type alias FileStats`](https://elm-doc-preview.netlify.app/BackendTask.Glob?repo=dillonkearns%2Felm-pages&version=master#FileStats) =
    { fullPath : String.String
    , sizeInBytes : Basics.Int
    , lastContentChange : Time.Posix
    , lastAccess : Time.Posix
    , lastFileChange : Time.Posix
    , createdAt : Time.Posix
    , isDirectory : Basics.Bool
    }
[`captureStats`](https://elm-doc-preview.netlify.app/BackendTask-Glob?repo=dillonkearns%2Felm-pages&version=master#captureStats) :
    BackendTask.Glob.Glob (BackendTask.Glob.FileStats -> value)
    -> BackendTask.Glob.Glob value
[`fromString`](https://elm-doc-preview.netlify.app/BackendTask-Glob?repo=dillonkearns%2Felm-pages&version=master#fromString) :
    String.String
    -> BackendTask.BackendTask error (List.List String.String)
[`fromStringWithOptions`](https://elm-doc-preview.netlify.app/BackendTask.Glob?repo=dillonkearns%2Felm-pages&version=master#fromStringWithOptions) :
    BackendTask.Glob.Options
    -> String.String
    -> BackendTask.BackendTask error (List.List String.String)

### Added

[`command`](https://elm-doc-preview.netlify.app/Pages-Script?repo=dillonkearns%2Felm-pages&version=master#command) :
    String.String
    -> List.List String.String
    -> BackendTask.BackendTask FatalError.FatalError String.String
[`doThen`](https://elm-doc-preview.netlify.app/Pages-Script?repo=dillonkearns%2Felm-pages&version=master#doThen) :
    BackendTask.BackendTask error value
    -> BackendTask.BackendTask error ()
    -> BackendTask.BackendTask error value
[`exec`](https://elm-doc-preview.netlify.app/Pages-Script?repo=dillonkearns%2Felm-pages&version=master#exec) :
    String.String
    -> List.List String.String
    -> BackendTask.BackendTask FatalError.FatalError ()
[`expectWhich`](https://elm-doc-preview.netlify.app/Pages-Script?repo=dillonkearns%2Felm-pages&version=master#expectWhich) :
    String.String
    -> BackendTask.BackendTask FatalError.FatalError String.String
[`question`](https://elm-doc-preview.netlify.app/Pages-Script?repo=dillonkearns%2Felm-pages&version=master#question) : String.String -> BackendTask.BackendTask error String.String
[`sleep`](https://elm-doc-preview.netlify.app/Pages-Script?repo=dillonkearns%2Felm-pages&version=master#sleep) : Basics.Int -> BackendTask.BackendTask error ()
[`which`](https://elm-doc-preview.netlify.app/Pages-Script?repo=dillonkearns%2Felm-pages&version=master#which) :
    String.String
    -> BackendTask.BackendTask error (Maybe.Maybe String.String)

**Comment by miniBill** [comment]
It's weird that `fileRead`'s error is `()`

**Reply by dillonkearns** [reply]
Yeah, I could have an error type for that, but it would only be accessible if it was the last part of the chain. So seems kind of niche, because at that point you may as well use the `BackendTask.File` APIs. So I'm not sure it's worth all the extra API surface area.

**Comment by miniBill** [comment]
The example in `Do.glob` says it calls `rm -rf` but it doesn't

**Reply by dillonkearns** [reply]
Fixed! Thank you for the feedback 🙏

**Comment by miniBill** [comment]
I still think `Do.allowFatal` makes sense

**Reply by dillonkearns** [reply]
Added! Thank you for the reminder 🙏
